### PR TITLE
[#3838]: Move to yarn instead of npm

### LINKED ISSUE #2541: npm install breaks lock files across developer machines
### User Story

As a developer, I want to have consistent package lock files so that I don't go crazy.

### Description

*Type*: Bug

*Summary*: `npm` is kind of busted–running `npm install` on my machine modifies the existing `package-lock.json` file. This is wrong behavior, as I have not added/removed/upgraded any packages.

#### Expected behavior

Running `npm install` out of the box does not swap around the lock file.

#### Actual behavior

Probably depending on your machine, the lock file will change. They should call it `package-lock-except-for-a-few-packages.json`

Here is the diff that gets produced on my machine:

```
diff --git i/package-lock.json w/package-lock.json
index 207e0166..6fef781f 100644
--- i/package-lock.json
+++ w/package-lock.json
@@ -446,12 +446,56 @@
           "resolved": "https://registry.npmjs.org/babel-register/-/babel-register-6.26.0.tgz",
           "integrity": "sha1-btAhFz4vy0htestFxgCahW9kcHE=",
           "requires": {
+            "babel-core": "6.26.0",
             "babel-runtime": "6.26.0",
             "core-js": "2.5.1",
             "home-or-tmp": "2.0.0",
             "lodash": "4.17.4",
             "mkdirp": "0.5.1",
             "source-map-support": "0.4.18"
+          },
+          "dependencies": {
+            "babel-core": {
+              "version": "6.26.0",
+              "resolved": "https://registry.npmjs.org/babel-core/-/babel-core-6.26.0.tgz",
+              "integrity": "sha1-rzL3izGm/O8RnIew/Y2XU/A6C7g=",
+              "requires": {
+                "babel-code-frame": "6.26.0",
+                "babel-generator": "6.26.0",
+                "babel-helpers": "6.24.1",
+                "babel-messages": "6.23.0",
+                "babel-register": "6.26.0",
+                "babel-runtime": "6.26.0",
+                "babel-template": "6.26.0",
+                "babel-traverse": "6.26.0",
+                "babel-types": "6.26.0",
+                "babylon": "6.18.0",
+                "convert-source-map": "1.5.0",
+                "debug": "2.6.9",
+                "json5": "0.5.1",
+                "lodash": "4.17.4",
+                "minimatch": "3.0.4",
+                "path-is-absolute": "1.0.1",
+                "private": "0.1.8",
+                "slash": "1.0.0",
+                "source-map": "0.5.7"
+              }
+            },
+            "babel-generator": {
+              "version": "6.26.0",
+              "resolved": "https://registry.npmjs.org/babel-generator/-/babel-generator-6.26.0.tgz",
+              "integrity": "sha1-rBriAHC3n248odMmlhMFN3TyDcU=",
+              "requires": {
+                "babel-messages": "6.23.0",
+                "babel-runtime": "6.26.0",
+                "babel-types": "6.26.0",
+                "detect-indent": "4.0.0",
+                "jsesc": "1.3.0",
+                "lodash": "4.17.4",
+                "source-map": "0.5.7",
+                "trim-right": "1.0.1"
+              }
+            }
           }
         }
       }
@@ -7001,9 +7045,9 @@
       }
     },
     "realm": {
-      "version": "2.0.3",
-      "resolved": "https://registry.npmjs.org/realm/-/realm-2.0.3.tgz",
-      "integrity": "sha1-F+vG4qBx9AaFZxdeQ6bIrFfQaT0=",
+      "version": "2.0.7",
+      "resolved": "https://registry.npmjs.org/realm/-/realm-2.0.7.tgz",
+      "integrity": "sha1-1SbWxzksVklafHgASDwG61hcbdY=",
       "requires": {
         "command-line-args": "4.0.7",
         "decompress": "4.2.0",
```

Either someone forgot to update `package-lock.json` when adding/upgrading some packages, or I'm seeing the same old `npm` behavior I'm used to.

### Reproduction

For me it was checking out the repo (OS X), and running `npm install` out of the box.

### Solution

Here are some links I tracked down today:

* https://github.com/npm/npm/issues/17722
* https://github.com/npm/npm/issues/18135
* https://github.com/npm/npm/issues/18712
* https://stackoverflow.com/questions/45022048/why-does-npm-install-rewrite-package-lock-json

To fix this on other projects, I've found the easiest and cheapest thing to do was to just switch to Yarn:

* Takes about 15 minutes.
* Yarn caches your builds in `~/.yarn/cache`, so future `yarn install` runs are really fast (~1-2 seconds if you haven't changed your packages)
* The lock files don't change until you tell them to

There's a bit of an extra cost here, which is adding one more dependency (`brew install yarn`). After that it's smooth sailing though!

### Additional Information

* Operating System: OS X building

### LINKED ISSUE #3838: Move to yarn instead of npm
[comment]: # (Please replace ... with your information. Remove < and >)

### User Story

As a developer,
I would like to use yarn instead of npm,
because it's faster (in our case) and does not get stuck installing packages

### Description

We agreed to move to yarn instead of npm.

We tried to do so but failed as we were not able to build a release in a reasonable amount of time.

Moving to yarn is a trivial code change, but after:
`lein prod-build`

and

`cd android && ./gradlew assembleRelease`

it would take a very long time to create the artifact, around 6 hrs on my local machine ( compared to 6 minutes when installing packages through npm). 

This has been replicated locally, on docker and on the macos build server.

The steps that fails is `app:bundleReleaseJsAndAssets`.

We see `jest-worker` hogging the CPU at 100%, memory consumption is also quite high, once it actually failed because the node engine could not allocate memory.

### A few scattered thoughts that might help

There are a few issues (some closed) https://github.com/facebook/react-native/issues/11696 which seems to indicate that it's bundling files that it should not be (realm files for example).

Some people mentioned `.babelrc` was the culprit, but did not elaborate any further.

### What we tried, but did not work

Increase node memory from 4GB to 8GB
Increase gradlew memory from 1GB to 4GB
Add `___tests__` to `inputExcludes` in `android/app/react.gradle`
Decrease the number of `jest-workers`

Currently we don't have the bandwidth to look into this any further, but if anyone wants to have a go at this it would be helpful.

### PR DESCRIPTION
**Pull request by alwx**
Fixes #3838.
Fixes #2541.

It also includes updates to `scripts/setup` — now it always installs the right version of node and yarn.

Checked it on my machine (MacBook 13" 2016, latest macOS) and on a new machine with freshly installed macOS (I literally installed macOS and immediately after that checked out Status and ran `scripts/setup`).

### CODE REVIEW

File: Jenkinsfile.nightly_fastlane [hunk 1]
```diff
@@ -11,7 +11,7 @@ def installJSDeps() {
     def installed = false
     while (!installed && attempt <= maxAttempts) {
         println "#${attempt} attempt to install npm deps"
-        sh 'npm install'
+        sh 'yarn install'
         installed = fileExists('node_modules/web3/index.js')
```
Comment (jeluard) [inline Jenkinsfile.nightly_fastlane]: Is this hack still needed?

File: ios/prebuild.log [hunk 2]
```diff
@@ -0,0 +1,2 @@
+Pre-action: patching react files!
```
Comment (dmitryn) [inline ios/prebuild.log]: Maybe add `ios/prebuild.log` to `.gitignore`?

File: env/dev/env/config.cljs [hunk 3]
```diff
@@ -1,6 +1,5 @@
 (ns env.config)
 
-(def figwheel-urls {
-                    :ios "ws://localhost:3449/figwheel-ws"
-                    :android "ws://10.0.3.2:3449/figwheel-ws"
-                    })
\ No newline at end of file
+(def figwheel-urls {:android "ws://192.168.10.203:3449/figwheel-ws",
+ :ios "ws://localhost:3449/figwheel-ws"}
```
Comment (dmitryn) [inline env/dev/env/config.cljs]: Formatting seems weird, that's how Cursive formats this:
```
(def figwheel-urls {:android "ws://192.168.0.104:3449/figwheel-ws"
                    :ios     "ws://localhost:3449/figwheel-ws"})
```

Comment (jeluard) [reply env/dev/env/config.cljs]: Do we need this change at all?

**Comment by dmitryn** [comment]
I also have `yarn-error.log` file generated which is not gitignored. Please gitignore it.

File: env/dev/env/config.cljs (hunk 3 above)
Comment (dmitryn) [reply env/dev/env/config.cljs]: Right, i usually don't commit this file.

**Comment by jeluard** [comment]
On linux I have the following message when running `react-native run-android` (then fails)

```
* What went wrong:
A problem occurred configuring project ':react-native-android'.
> Could not get unknown property 'repositoryUrl' for project ':react-native-android' of type org.gradle.api.Project.
```

But not sure if it is related to those changes.

**Comment by dmitryn** [comment]
@jeluard i have the same error on OSX for `make run-android`:

```
* What went wrong:
A problem occurred configuring project ':react-native-android'.
> Could not get unknown property 'repositoryUrl' for project ':react-native-android' of type org.gradle.api.Project.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

* Get more help at https://help.gradle.org

BUILD FAILED in 10s
make: *** [run-android] Error 1
```

**Comment by dmitryn** [comment]
`make run-ios` showing the build error:

```
In file included from /Users/trybeee/Code/status-react/node_modules/react-native/third-party/glog-0.3.4/src/symbolize.cc:55:
In file included from /Users/trybeee/Code/status-react/node_modules/react-native/React/../third-party/glog-0.3.4/src/utilities.h:73:
/Users/trybeee/Code/status-react/node_modules/react-native/React/../third-party/glog-0.3.4/src/base/mutex.h:105:10: fatal error: 'config.h' file not found
#include "config.h"           // to figure out pthreads support
         ^~~~~~~~~~


** BUILD FAILED **



The following build commands failed:
        CompileC /Users/trybeee/Code/status-react/ios/build/Build/Intermediates.noindex/React.build/Debug-iphonesimulator/third-party.build/Objects-normal/x86_64/raw_logging.o /Users/trybeee/Code/status-react/node_modules/react-native/third-party/glog-0.3.4/src/raw_logging.cc normal x86_64 c++ com.apple.compilers.llvm.clang.1_0.compiler
        CompileC /Users/trybeee/Code/status-react/ios/build/Build/Intermediates.noindex/React.build/Debug-iphonesimulator/third-party.build/Objects-normal/x86_64/utilities.o /Users/trybeee/Code/status-react/node_modules/react-native/third-party/glog-0.3.4/src/utilities.cc normal x86_64 c++ com.apple.compilers.llvm.clang.1_0.compiler
(2 failures)

Installing build/Build/Products/Debug-iphonesimulator/StatusIm.app
Launching im.status.ethereum.debug
im.status.ethereum.debug: 43232
```

but then starts app and it works well there.

**Comment by dmitryn** [comment]
@jeluard what gradle version do you have? I have `Gradle 4.8.1`
Seems like related to this issue https://stackoverflow.com/questions/43967489/could-not-get-unknown-property-repositoryurl-for-project

<blockquote><div>Stack Overflow</div><div><strong><a href="https://stackoverflow.com/questions/43967489/could-not-get-unknown-property-repositoryurl-for-project">Could not get unknown property 'repositoryUrl' for project</a></strong></div><div>I have a project employing React-Native. I'm building the RN module from source, so my project has the ReactAndroid module as a dependency.

I was trying to upgrade the project to Android gradle bu...</div></blockquote>

**Comment by jeluard** [comment]
@dmitryn No idea about my gradle version

**Comment by dmitryn** [comment]
@jeluard added these changes to `node_modules/react-native/ReactAndroid/release.gradle` : 
```
def getRepositoryUrl() {
    return findProperty('repositoryUrl') != null ? property('repositoryUrl') : 'https://oss.sonatype.org/service/local/staging/deploy/maven2/'
}

def getRepositoryUsername() {
    return findProperty('repositoryUsername') !=null ? property('repositoryUsername') : ''
}

def getRepositoryPassword() {
    return findProperty('repositoryPassword') != null ? property('repositoryPassword') : ''
}
```
and `make run-android` works well now.

Not sure how this error is related to switch to Yarn though.

**Comment by dmitryn** [comment]
Also, Jenkins build is failing with:
```

[react_pull_requests_PR-5056-PLNJRE3W6MYRZC36I7TMFKLT6JKWFLHCO4OYANWJOEXVVVKVIMWQ] Running shell script

+ yarn install

yarn install v1.3.2

[1/4] Resolving packages...

warning react-native > connect@2.30.2: connect 2.x series is deprecated

[2/4] Fetching packages...

warning Pattern ["bignumber.js@github:status-im/bignumber.js#master"] is trying to unpack in the same destination "/Users/jenkins/Library/Caches/Yarn/v1/npm-bignumber.js-4.0.2-cc066a0a3d6bfe0c436c9957f4ea8344bf963c89" as pattern ["bignumber.js@git+https://git@github.com/status-im/bignumber.js.git"]. This could result in a non deterministic behavior, skipping.

[3/4] Linking dependencies...

warning " > react-native-crypto@2.1.1" has unmet peer dependency "react-native-randombytes@>=2.0.0 <4.0.0".

warning " > react-native-level-fs@3.0.0" has incorrect peer dependency "asyncstorage-down@^3.0.0".

warning "rn-snoopy > eslint-plugin-jsx-a11y@5.1.1" has unmet peer dependency "eslint@^2.10.2 || ^3 || ^4".

error An unexpected error occurred: "ENOENT: no such file or directory, open '/Users/jenkins/Library/Caches/Yarn/v1/npm-bignumber.js-4.0.2-9face0f38fadaa3955348af926e8960a14b79346/.yarn-metadata.json'".

info If you think this is a bug, please open a bug report with the information provided in "/Users/jenkins/workspace/react_pull_requests_PR-5056-PLNJRE3W6MYRZC36I7TMFKLT6JKWFLHCO4OYANWJOEXVVVKVIMWQ/yarn-error.log".

info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.

script returned exit code 1
```

**Comment by rasom** [comment]
maybe we have wrong version of yarn on jenkins or something like this, @alwx which version did you use?

**Review by jeluard — APPROVED** [review]

**Comment by jeluard** [comment]
@dmitryn Could you review this PR again?

**Review by dmitryn — APPROVED** [review]
Works fine on my machine. Jenkins build is failing for some reason though.

**Comment by alwx** [comment]
Checking.

**Comment by alwx** [comment]
If only I were able to access Jenkins. It's down now :(

File: scripts/lib/setup/installers.sh [hunk 4]
```diff
@@ -176,17 +176,17 @@ function install_maven() {
...
-    npm_command="sudo npm"
+    yarn_command="sudo yarn"
   fi
 
-  if npm list -g | grep -q react-native-cli; then
+  if yarn global list | grep react-native-cli; then
     already_installed "react-native-cli"
   else
-    $npm_command install -g react-native-cli
+    $yarn_command global install react-native-cli
```
Comment (jeluard) [inline scripts/lib/setup/installers.sh]: This should be `yarn_command global add`

**Review by jeluard — CHANGES_REQUESTED** [review]

**Comment by jeluard** [comment]
Can we remove `re-natal` symlinks now? Fails for me.

**Comment by cammellos** [comment]
@alwx seems like we hit the same problem as in #3838 

https://jenkins.status.im/blue/organizations/jenkins/status-react%2Fpull%20requests/detail/PR-5056/9/pipeline is stuck on `:app:bundleReleaseJsAndAssets`, any idea on how to fix or what might be the cause?

<blockquote><div><strong><a href="https://jenkins.status.im/blue/organizations/jenkins/status-react%2Fpull%20requests/detail/PR-5056/9/pipeline">Jenkins</a></strong></div></blockquote>

**Comment by goranjovic** [comment]
@alwx should we close this? or is it still possible to merge?

**Comment by jakubgs** [comment]
All of those `Jenkinsfile` changes are pointless since we don't use those anymore. The only file you should be worrid about is this:
https://github.com/status-im/status-react/blob/develop/ci/common.groovy#L48

The rest should be removed, I even have a PR for that:
https://github.com/status-im/status-react/pull/5710

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/status-im/status-react">status-im/status-react</a></strong></div><div>a free (libre) open source, mobile OS for Ethereum - status-im/status-react</div></blockquote>

**Comment by jakubgs** [comment]
Okay, old unused `Jenkinsfile`s have been removed in https://github.com/status-im/status-react/commit/495f513de9c108f9191e5740e73bc0f70f1a2778. Please rebase this branch.

**Comment by jakubgs** [comment]
I'd really like to see this merged, because `npm` is causing a lot of various problems. What does this need?

**Comment by jakubgs** [comment]
I played with this a bit but it seems like the failures require someone more well versed in the codebase itself.

**Comment by jeluard** [comment]
Closing as discussed (long ago 😅 ) Still too many technical constraints to move on to yarn.

**Comment by jeluard** [comment]
@alwx Please add any relevant details !

**Comment by gnl** [comment]
Reopening and taking over as discussed with @oskarth .

**Comment by oskarth** [comment]
@rasom Why did you close this?

**Comment by rasom** [comment]
i suppose the issue is still open and we will return to this, but PR hasn't been touched for 29 days

**Comment by oskarth** [comment]
Oh sorry you are right, I thought I was in an issue, not a PR. My bad :)

**Pull request closed without merging**